### PR TITLE
added getResourceAccountAddress function along with unit tests

### PR DESCRIPTION
**Pull request by dhruvja**
### Description

This PR is related to the issue  #3739  created to add a function in TS SDK to fetch the resource account given that the source address and seeds are provided.

### Test Plan

I have added a function `getResourceAccountAddress` with 2 parameters `sourceAddress` and `seeds` which are used to derive the resource account. Since the resource account derivation is expensive on chain, we can use this method to derive if off chain and send it to the smart

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3822)

### CODE REVIEW

File: ecosystem/typescript/sdk/src/aptos_account.test.ts [hunk 1]
```diff
@@ -65,3 +65,12 @@ test("Signs Strings", () => {
     "0xc5de9e40ac00b371cd83b1c197fa5b665b7449b33cd3cdd305bb78222e06a671a49625ab9aea8a039d4bb70e275768084d62b094bc1b31964f2357b7c1af7e0d",
   );
 });
+
+test("Gets the resource account address", () => {
+  const sourceAddress = "9db1d65a321e4a86f8098cb04e76fce098890b84211fa06d65f24dc644bf0fec";
+  const seeds = "3031" // hexadecimal form of "01" which is 3031
```
Comment (davidiw) [inline ecosystem/typescript/sdk/src/aptos_account.test.ts:71]: nit seed

File: ecosystem/typescript/sdk/src/aptos_account.test.ts [hunk 2]
```diff
@@ -65,3 +65,12 @@ test("Signs Strings", () => {
...
+
+  const a1 = new AptosAccount();
+
+  expect(a1.getResourceAccountAddress(sourceAddress, seeds).hex()).toBe("0x915e47f986471a5faba9cb2f726611cd554344419ddd6d5b3e9f00bcafd30169");
```
Comment (davidiw) [inline ecosystem/typescript/sdk/src/aptos_account.test.ts:75]: need seeds

**Review by davidiw — COMMENTED** [review]
address the lints, nits, and let's not update the parckage-lock or yarn.lock

thanks!

File: ecosystem/typescript/sdk/src/aptos_account.test.ts (hunk 2 above)
Comment (dhruvja) [reply ecosystem/typescript/sdk/src/aptos_account.test.ts]: I didnt understand, what did u mean?

**Review by davidiw — COMMENTED** [review]
Please drop the package-lock.json and yarn.lock and we can land.

File: ecosystem/typescript/sdk/src/aptos_account.ts [hunk 3]
```diff
@@ -140,6 +142,29 @@ export class AptosAccount {
...
+    const seed = new HexString(seeds).toUint8Array();
+
+    const originBytes = new Uint8Array(source.length + seed.length);
+    this.address();
+
+    originBytes.set(source);
+    originBytes.set(seed, source.length);
+
+    const hash = SHA3.sha3_256.create();
+    hash.update(Buffer.from(originBytes));
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/aptos_account.ts:163]: Don't use `Buffer` here. It doesn't work in browsers.

File: ecosystem/typescript/sdk/src/aptos_account.ts [hunk 4]
```diff
@@ -140,6 +142,29 @@ export class AptosAccount {
...
+   * @param seeds The seeds need to be in hexadecimal format
+   * @returns The resource account address
+   */
+
+  getResourceAccountAddress(sourceAddress: MaybeHexString, seeds: string): HexString {
+    const source = bcsToBytes(AccountAddress.fromHex(sourceAddress));
+    const seed = new HexString(seeds).toUint8Array();
+
+    const originBytes = new Uint8Array(source.length + seed.length);
+    this.address();
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/aptos_account.ts:157]: huh?

File: ecosystem/typescript/sdk/src/aptos_account.ts [hunk 5]
```diff
@@ -140,6 +142,29 @@ export class AptosAccount {
...
   }
 
+  /**
+   * Takes source address and seeds and returns the resource account address
+   * @param sourceAddress Address used to derive the resource account
+   * @param seeds The seeds need to be in hexadecimal format
+   * @returns The resource account address
+   */
+
+  getResourceAccountAddress(sourceAddress: MaybeHexString, seeds: string): HexString {
```
Comment (jjleng) [inline ecosystem/typescript/sdk/src/aptos_account.ts:152]: Should be a static method?

Comment (jjleng) [reply ecosystem/typescript/sdk/src/aptos_account.ts:152]: Also, it is better to make seeds Uint8Array

Comment (dhruvja) [reply ecosystem/typescript/sdk/src/aptos_account.ts:152]: Yes it should be a static method.

Comment (dhruvja) [reply ecosystem/typescript/sdk/src/aptos_account.ts:152]: ok ill make it

Comment (dhruvja) [reply ecosystem/typescript/sdk/src/aptos_account.ts:152]: > Also, it is better to make seeds Uint8Array

Ill make seeds Uint8Array

File: ecosystem/typescript/sdk/src/aptos_account.ts (hunk 4 above)
Comment (dhruvja) [reply ecosystem/typescript/sdk/src/aptos_account.ts:157]: if we make it a static method, we can remove `this.address()`. Sorry my bad.

File: ecosystem/typescript/sdk/src/aptos_account.ts (hunk 3 above)
Comment (dhruvja) [reply ecosystem/typescript/sdk/src/aptos_account.ts]: Done

**Review by jjleng — APPROVED** [review]

**Comment by jjleng** [comment]
`yarn.lock` should be untouched. You deleted it. Please fix that

**Comment by dhruvja** [comment]
> `yarn.lock` should be untouched. You deleted it. Please fix that

i have added it

**Comment by jjleng** [comment]
see https://github.com/aptos-labs/aptos-core/pull/4173. I would table this PR until the smart contract side change is in.

**Comment by davidiw** [comment]
Please update and use reset the yarn.lock to the previous commit.

**Comment by dhruvja** [comment]
> Please update and use reset the yarn.lock to the previous commit.

I have updated yarn.lock to the previous commit

**Comment by davidiw** [comment]
this is very puzzling, are you unable to fix the yarn.lock issue?
feel free to re-open if you can, the PR is hard to track

**Pull request closed without merging**